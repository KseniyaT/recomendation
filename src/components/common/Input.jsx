import React from 'react';
import PropTypes from 'prop-types';

function Input(props) { // eslint-disable-line require-jsdoc
  return (
    <div className="input-group">
      <div className="input-group__input-container">
        <input
          type={props.type}
          id={props.id}
          name={props.name}
          className="input-group__input"
          onChange={props.onChange}
          placeholder={props.placeholder}
        />
      </div>
    </div>
  );
}

Input.defaultProps = {
  type: 'text',
};

Input.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  onChange: PropTypes.func,
};

export default Input;
