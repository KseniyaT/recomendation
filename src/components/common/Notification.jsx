import React from 'react';
import PropTypes from 'prop-types';

function Notification(props) { // eslint-disable-line require-jsdoc
  return (
    <div className="notification notification_error">{props.errorMsg}</div>
  );
}

Notification.propTypes = {
  errorMsg: PropTypes.string,
};

export default Notification;
