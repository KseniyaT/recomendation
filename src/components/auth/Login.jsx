import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import actions from '../../actions/auth/index';
import Input from '../common/Input';
import Notification from '../common/Notification';

const { loginUser } = actions;

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      password: '',
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   * It saves users printed values in an input to according state of the component
   * @param {Event} event - event object
   * @returns {void}
   */
  handleInputChange(event) {
    this.setState({
      [event.target.id]: event.target.value,
    });
  };

  /**
   * It requested server to login an user with username/password
   * @param {Event} event - event object
   * @returns {void}
   */
  handleSubmit(event) {
    event.preventDefault();
    this.props.loginUser(this.state);
  };

  render() {
    const { errorMsg } = this.props;
    return (
      <form className="form form_centred" onSubmit={this.handleSubmit}>
        <h1 className="title">Login</h1>
        {
          errorMsg && <Notification errorMsg={errorMsg} />
        }
        <Input
          id="username"
          name="username"
          type="text"
          placeholder="Username"
          onChange={this.handleInputChange}
        />
        <Input
          id="password"
          name="password"
          type="password"
          placeholder="Password"
          onChange={this.handleInputChange}
        />
        <div className="input-group">
          <button className="btn btn_full" type="submit">Login</button>
        </div>
      </form>
    );
  }
}

Login.propTypes = {
  errorMsg: PropTypes.string,
  loginUser: PropTypes.func,
};

const mapStateToProps = (state) => ({
  errorMsg: state.auth.errorMsg,
});

const mapDispatchToProps = (dispatch) => ({
  loginUser: (userInfo) => dispatch(loginUser(userInfo)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
