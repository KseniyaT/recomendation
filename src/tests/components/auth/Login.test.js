import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Login from '../../../components/auth/Login';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import Input from '../../../components/common/Input';
import Notification from '../../../components/common/Notification';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Login', () => {
  let store;
  let rendererComponent;
  let wrapper;

  beforeEach(() => {
    store = mockStore({
      auth: {
        errorMsg: '',
      },
    });

    store.dispatch = jest.fn();

    rendererComponent = renderer.create(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    wrapper = mount( <Provider store={store}> <Login /> </Provider> );
  });

  it('should render with given state from Redux store', () => {
    expect(rendererComponent.toJSON()).toMatchSnapshot();
  });

  it('should call dispatch method when the form is submitted', () => {
    const fakeEvent = { preventDefault: () => {} };
    rendererComponent.root.findByType('form').props.onSubmit(fakeEvent);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('renders main rendererComponents', () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper.find(Input).length).toBe(2);
    expect(wrapper.find('button').length).toBe(1);
  });

  it('changes corresponding states when inputs values are changed', () => {
    const wrapperClass = mount(shallow(<Login store={ store } />).get(0));
    wrapperClass.find('input[type="text"]').simulate('change', { target: { id: 'username', value: 'usernametest' }});
    expect(wrapperClass.state('username')).toEqual('usernametest');
    wrapperClass.find('input[type="password"]').simulate('change', { target: { id: 'password', value: 'Qwerty' }});
    expect(wrapperClass.state('password')).toEqual('Qwerty');
  });

  it('should show an error message then it is and should not when the errorMsg is empty', () => {
    const store1 = mockStore({
      auth: {
        errorMsg: 'Message',
      },
    });
    const containerWithOutError = mount(<Login store={ store } />);
    expect(containerWithOutError.find(Notification).length).toBe(0);
    const containerWithError = mount(<Login store={ store1 } />);
    expect(containerWithError.find(Notification).length).toBe(1);
    expect(containerWithError.find(Notification).text()).toEqual('Message');
  });

});
