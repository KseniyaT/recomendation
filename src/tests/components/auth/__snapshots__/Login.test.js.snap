// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render with given state from Redux store 1`] = `
<form
  className="form form_centred"
  onSubmit={[Function]}
>
  <h1
    className="title"
  >
    Login
  </h1>
  
  <div
    className="input-group"
  >
    <div
      className="input-group__input-container"
    >
      <input
        className="input-group__input"
        id="username"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        type="text"
      />
    </div>
  </div>
  <div
    className="input-group"
  >
    <div
      className="input-group__input-container"
    >
      <input
        className="input-group__input"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </div>
  </div>
  <div
    className="input-group"
  >
    <button
      className="btn btn_full"
      type="submit"
    >
      Login
    </button>
  </div>
</form>
`;
